pkg/services/ngalert/state/cache.go

```diff
@@ -224,6 +223,25 @@ func (c *cache) recordMetrics() {
 	}
 }
 
+func buildTemplateData(result eval.Result) (map[string]string, map[string]string) {
+	labels := make(map[string]string)
+	for k, v := range result.Instance {
+		labels[k] = v
+	}
+	values := make(map[string]string)
+	for k, v := range result.Values {
```

Comment:
It might be better if this is an actual float64 (or if we go with the struct I mentioned earlier, the property would be float), I am think this might be the case so people can do number formatting with the template built-in `printf`

Reply:
I think it would need to be `*float64` since the value can be absent? I'm not sure I want to use `0` because `0` is a valid value and in the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

Reply:
I think it would need to be `*float64` since the value can be absent?

I don't think I want to use `0` because `0` is a valid metric value. In the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

Reply:
I think it would need to be `*float64` since the value can be absent? I don't think I want to use `0` because `0` is a valid metric value. In the value string we use `"null"` in places of absent values. If we use `*float64` do we need to be concerned about templates dereferencing nil?

Reply:
Yes pointer seems right. It can be nil, because {{ pipeline }} is calling printf

Example to show it doesn't panic: https://play.golang.org/p/_LmX3iciRG9

Reply:
Oh wait no I'm wrong, will print address the value... Wonder how prom does the number formating ...

Reply:
Looks like `humanzie` https://prometheus.io/docs/prometheus/latest/configuration/template_reference/#numbers ... not sure what is best here https://github.com/JensErat/prometheus/blob/master/template/template.go#L167

Float64 non-pointer is maybe fine if we create the container, as I expect that the entry won't exist if there wasn't a number for it so the containing object wouldn't exist?